src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -1204,8 +1204,51 @@ private TokenValidationResult ValidateJWS(string token, TokenValidationParameter
         {
             try
             {
-                var jsonWebToken = ValidateSignature(token, validationParameters, configuration);
-                return ValidateTokenPayload(jsonWebToken, validationParameters, configuration);
+                JsonWebToken jsonWebToken;
+                TokenValidationResult tokenValidationResult;
+                if (validationParameters.SignatureValidator != null || validationParameters.SignatureValidatorUsingConfiguration != null)
+                {
+                    return ValidateTokenPayload(
+                        ValidateSignatureUsingDelegates(token, validationParameters, configuration),
+                        validationParameters,
+                        configuration);
+                }
+                else
+                {
+                    if (validationParameters.TokenReader != null)
+                    {
+                        SecurityToken securityToken = validationParameters.TokenReader(token, validationParameters);
+                        if (securityToken == null)
+                            throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, token)));
+
+                        jsonWebToken = securityToken as JsonWebToken;
+                        if (jsonWebToken == null)
+                            throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10509, LogHelper.MarkAsNonPII(typeof(JsonWebToken)), LogHelper.MarkAsNonPII(securityToken.GetType()), token)));
+                    }
+                    else
+                    {
+                        jsonWebToken = new JsonWebToken(token);
+                    }
+
+                    if (validationParameters.ValidateSignatureLast)
+                    {
+                        tokenValidationResult = ValidateTokenPayload(jsonWebToken, validationParameters, configuration);
+                        if (tokenValidationResult.IsValid)
+                        {
+                            tokenValidationResult.SecurityToken = ValidateSignature(token, jsonWebToken, validationParameters, configuration);
+                            Validators.ValidateIssuerSecurityKey(tokenValidationResult.SecurityToken.SigningKey, jsonWebToken, validationParameters);
```

Comment:
Should ValidateSignature and ValidateIssuerSecurityKey be coupled? so that we can avoid checking the flag in ValidateTokenPayload.

Reply:
I think Roja's suggestion makes sense. Having to check the `validationParameters.ValidateSignatureLast` in two methods more error prone. Consider wrapping the two methods:

```
        private static JsonWebToken ValidateSignatureAndIssuerSecurityKey(string token, JsonWebToken jsonWebToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
        {
            var validatedToken = ValidateSignature(token, jsonWebToken, validationParameters, configuration);
            Validators.ValidateIssuerSecurityKey(validatedToken.SigningKey, jsonWebToken, validationParameters);

            return validatedToken;
        }
```